Improve cmake config

ericpre opened the pull request:
A couple of improvements and fixes to build prismatic on conda-forge: https://github.com/conda-forge/staged-recipes/pull/9834, https://github.com/conda-forge/prismatic_split-feedstock/pull/5 and https://github.com/conda-forge/prismatic_split-feedstock/pull/6.

- Add option to set output name for cmake (useful to make an "prismatic-double" executable with double precision, for example).
- Find hdf5 on windows.
- Update findFFTW.cmake (from https://github.com/egpbos/findFFTW) and fix windows build.
- Don't link against libpython on macOS to avoid segmentation fault when importing pyprismatic. Set pyprismatic_core lib extension to "so" instead of "dylib" on macos.

@lerandc @sezelt @douglowe since you have been discussing building pyprismatic on macos, you may be interesting in the fix suggested in this PR.

lerandc commented:
It all looks good to me, thanks for helping with these changes!

ericpre commented:
@lerandc: thanks for looking a this!

sezelt commented on CMakeLists.txt:
```diff
@@ -80,21 +101,21 @@ endif(PRISMATIC_ENABLE_PYPRISMATIC)
    message("entered")
    set(HDF5_USE_STATIC_LIBRARIES ON)
+   set (SEARCH_TYPE static)
++else(NOT PRISMATIC_USE_HDF5_STATIC)
```
this `+` appears to be a typo

sezelt commented:
hey @ericpre this looks really great! But I am having trouble building on my Mac.

I don't think this has to do with your changes, rather just my build environment, but I am stuck. Any chance you know what to do?

At first I could not get cmake to find HDF5, even though I have several installs and tried pointing it to all of them. Then I realized that the binary dists of HDF5 that I had did not have a cmake file included, so I built it from source (using ctest as per HDF website). Now when I try to build Prismatic I get this cmake error

```
CMake Error at /Users/sez/work/hdf5/CMake-hdf5-1.10.1/hdf5Config.cmake:262 (include):
  include could not find load file:

    /HDF5options.cmake
Call Stack (most recent call first):
  CMakeLists.txt:111 (find_package)


CMake Error at /Users/sez/work/hdf5/CMake-hdf5-1.10.1/hdf5Config.cmake:263 (include):
  include could not find load file:

    /CTestScript.cmake
Call Stack (most recent call first):
  CMakeLists.txt:111 (find_package)


Boost dir /opt/local/include
FFTW dir /usr/local/include
Single precision enabled
-- Configuring incomplete, errors occurred!
See also "/Users/sez/git-repos/prismatic/build/CMakeFiles/CMakeOutput.log".
CMake Error: CTEST_USE_LAUNCHERS is enabled, but the RULE_LAUNCH_COMPILE global property is not defined.
Did you forget to include(CTest) in the toplevel CMakeLists.txt ?
```

I checked that HDF5options.cmake and CTestScript.cmake are in the directory /Users/sez/work/hdf5/CMake-hdf5-1.10.1 (which I have exported as $HDF5_ROOT). I also tried adding `include(CTest)` to the Prismatic `CMakeLists.txt` file and the error is the same.

Any ideas?

ericpre commented:
@sezelt, I will come back to this soon - I also had one issue with finding the hdf5 libraries with this branch.
In the mean time, the pyprismatic package for macos is now available on conda-forge, could you try that maybe?

ericpre commented:
@sezelt: I reverted (unintentional) changes in finding the hdf5 libraries, so it should work better now.

I have added CI configuration scripts for azure pipeline to build all and run the simulation demo. The log is accessible at https://dev.azure.com/ericpre/prismatic/_build/results?buildId=390.

These scripts are generated and maintain automatically using the conda-forge smithy tool: https://conda-forge.org/docs/user/ci-skeleton.html. Obviously, this is useful to check that the code build and runs fine.

lerandc commented:
@sezelt were you ever able to test @ericpre 's most recent updates on your system?

sezelt commented:
@lerandc actually something is so badly wrong with my build system that I can't compile any branch of prismatic anymore 🤯

ericpre commented:
Would it be sensible to setup azure pipeline for this repository? As mentioned in https://github.com/prism-em/prismatic/pull/76#issuecomment-613642174, this would check that the code builds fine on *all* platform and for *most* variants - GPU are build only on linux and there is test run, since there are no GPU accessible, but all other variant are being built (cli, gui, double precision and python binding). And once the test suite is being implemented, it will be ready to go!

lerandc commented:
Looking into it briefly, it seems like that could be a great solution for build checking. How hard is it to set up/maintain?

ericpre commented:
Regarding the build recipe, most of the hard work has been done in https://github.com/conda-forge/prismatic_split-feedstock and the recipe maintenance should be done in the feedstock repository. Once the conda build recipe has been setup properly, the maintenance work is very low - almost everything automated on conda-forge.
The primary purpose of conda-forge is to build packages but it can be used to build source code in CI too, in which case, we simply need to run conda-smithy to update the CI scripts (https://conda-forge.org/docs/user/ci-skeleton.html#keeping-up-to-date).
The only difference from https://github.com/conda-forge/prismatic_split-feedstock is where the source is pull from (in our case, locally): https://github.com/prism-em/prismatic/pull/76/files#diff-e178b687b10a71a3348107ae3154e44cR13-R18.

ericpre commented:
And to set up azure pipeline, there are instructions only but it can be a bit confusing. I have tried to summarise it: https://github.com/hyperspy/ci-scripts#setting-up-ci-services.

ericpre commented:
@lerandc: should I change the base to `dev` as for the other PRs?

thomasaarholt commented:
> @lerandc: should I change the base to `dev` as for the other PRs?

If you do, I will test this immediately (the cmake aspect). I was just about to compile prismatic.

ericpre commented:
> > @lerandc: should I change the base to `dev` as for the other PRs?
> 
> If you do, I will test this immediately (the cmake aspect). I was just about to compile prismatic.

This was referring to changing the `base` of the PR as in what branch it would be merged in. It would not change anything to this branch.

thomasaarholt commented:
> This was referring to changing the `base` of the PR as in what branch it would be merged in. It would not change anything to this branch.

I'm aware, I just couldn't recall how to merge a PR into the dev branch of my local repo. I am intending to compile the dev branch.

ericpre commented:
Ping @lerandc! ;)

lerandc commented:
Thanks for the ping, sorry for the late response!

ericpre commented:
This PR should have been closed automatically when it was merged (https://github.com/prism-em/prismatic/commit/9fdc88b8715f2b29b79fbf40ebdad517612c28b6) but it seems that github was confused...
